selfservice/strategy/code/strategy_login_test.go

```diff
@@ -0,0 +1,353 @@
+
+		i.VerifiableAddresses = va
+
+		require.NoError(t, reg.PrivilegedIdentityPool().CreateIdentity(ctx, i))
+		return i
+	}
+
+	type state struct {
+		flowID        string
+		csrfToken     string
+		identity      *identity.Identity
+		client        *http.Client
+		loginCode     string
+		identityEmail string
+	}
+
+	createLoginFlow := func(t *testing.T, moreIdentifiers ...string) *state {
+		t.Helper()
+
+		identity := createIdentity(t, moreIdentifiers...)
+		client := testhelpers.NewClientWithCookies(t)
+
+		// 1. Initiate flow
+		resp, err := client.Get(public.URL + login.RouteInitBrowserFlow)
+		require.NoError(t, err)
+
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		flowID := gjson.GetBytes(body, "id").String()
+		require.NotEmpty(t, flowID)
+
+		csrfToken := gjson.GetBytes(body, "ui.nodes.#(attributes.name==csrf_token).attributes.value").String()
+		require.NotEmpty(t, csrfToken)
+
+		require.NoError(t, resp.Body.Close())
+
+		loginEmail := gjson.Get(identity.Traits.String(), "email").String()
+		require.NotEmpty(t, loginEmail)
+
+		return &state{
+			flowID:        flowID,
+			csrfToken:     csrfToken,
+			identity:      identity,
+			identityEmail: loginEmail,
+			client:        client,
+		}
+	}
+
+	type onSubmitAssertion func(t *testing.T, s *state, res *http.Response)
+
+	submitLoginID := func(t *testing.T, s *state, submitAssertion onSubmitAssertion) *state {
+		t.Helper()
+
+		payload := strings.NewReader(url.Values{
+			"csrf_token": {s.csrfToken},
+			"method":     {"code"},
+			"identifier": {s.identityEmail},
+		}.Encode())
+
+		req, err := http.NewRequestWithContext(ctx, "POST", public.URL+login.RouteSubmitFlow+"?flow="+s.flowID, payload)
+		require.NoError(t, err)
+
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		req.Header.Set("Accept", "application/json")
+
+		resp, err := s.client.Do(req)
+		require.NoError(t, err)
+
+		if submitAssertion != nil {
+			submitAssertion(t, s, resp)
+			return s
+		}
+
+		require.EqualValues(t, http.StatusOK, resp.StatusCode)
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		csrfToken := gjson.GetBytes(body, "ui.nodes.#(attributes.name==csrf_token).attributes.value").String()
+		require.NotEmpty(t, csrfToken)
+
+		s.csrfToken = csrfToken
+
+		require.NoError(t, resp.Body.Close())
+
+		return s
+	}
+
+	submitLoginCode := func(t *testing.T, s *state, submitAssertion onSubmitAssertion) *state {
+		t.Helper()
+
+		req, err := http.NewRequestWithContext(ctx, "POST", public.URL+login.RouteSubmitFlow+"?flow="+s.flowID, strings.NewReader(url.Values{
+			"csrf_token": {s.csrfToken},
+			"method":     {"code"},
+			"code":       {s.loginCode},
+			"identifier": {s.identityEmail},
+		}.Encode()))
+		require.NoError(t, err)
+
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+		req.Header.Set("Accept", "application/json")
```

Comment:
Take a look at the password tests - we are testing various flows there (browser html, browser ajax, native api) and I think we need to do this here and in registration too. There are also some testhelpers you can use to easily submit these flows!